### PR TITLE
Email share tokens

### PR DESCRIPTION
**Pull request by que-sara-sara**

### CODE REVIEW

File: api/user.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+const firebase = require('firebase');
```
Comment (que-sara-sara) [inline api/user.js:1]: these are actually new things that i added

File: api/albums-controller.js [hunk 2]
```diff
@@ -48,14 +53,24 @@ async function joinAlbum(userToken, userId, shareToken) {
   return result;
 }
 
+async function joinPendingAlbums(userToken, userId) {
```
Comment (que-sara-sara) [inline api/albums-controller.js:65]: check it out

File: api/albums-router.js [hunk 3]
```diff
@@ -18,70 +21,72 @@ router.use((req, res, next) => {
...
     res.json({ album });
   } catch (err) {
-    console.log('ERRRRRRR', err);
+    console.log("ERRRRRRR", err);
     res.status(500).json(err);
   }
 });
 
-router.get('/:albumId/share-token', async(req, res) => {
+router.post("/:albumId/share-token", async (req, res) => {
```
Comment (que-sara-sara) [inline api/albums-router.js:33]: i changed to this to a post

File: api/albums-router.js [hunk 4]
```diff
@@ -18,70 +21,72 @@ router.use((req, res, next) => {
...
   const { albumId } = req.params;
   const userToken = req.user.token;
   const userId = req.user.profile.id;
+  const shareUser = req.body.shareUser;
   try {
-    const album = await shareAlbum(userToken, userId, albumId);
+    const album = await shareAlbum(userToken, userId, albumId, shareUser);
```
Comment (que-sara-sara) [inline api/albums-router.js:39]: `shareUser` should be an email address but we can discuss other identifier options

File: api/index.js [hunk 5]
```diff
@@ -1,46 +1,56 @@
...
   new OAuth2Strategy(
     oAuthConfig,
     // TODO: save the user to the database in this callback
-    (token, refreshToken, profile, done) => done(null, { profile, token })
+    async (token, refreshToken, profile, done) => {
+      console.log("token", token);
+      console.log("refreshToken", refreshToken);
+      console.log("profile", profile);
+
+      await joinPendingAlbums(token, profile.id);
```
Comment (que-sara-sara) [inline api/index.js:34]: this is new

File: src/index.js [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
 ReactDOM.render(
   <BrowserRouter>
     <App auth={auth} />
   </BrowserRouter>,
-  document.getElementById('root')
+  document.getElementById("root")
 );
 
-// register();
+serviceWorker.unregister();
```
Comment (kenzable) [inline src/index.js]: Why unregistering this? Will we end up needing to activate this later?

File: api/index.js [hunk 7]
```diff
@@ -1,46 +1,56 @@
...
+const session = require("express-session");
+const bodyParser = require("body-parser");
+const sessionFileStore = require("session-file-store");
+const passport = require("passport");
+const socketio = require("socket.io");
+const { oAuth: oAuthConfig, session: sessionConfig } = require("../config");
+const { initializeCache, joinPendingAlbums } = require("./albums/data-access");
+const authRouter = require("./user/auth");
+const albumsRouter = require("./albums/albums");
+const firebase = require("firebase");
```
Comment (kenzable) [inline api/index.js:10]: nitpick, should be moved up to before the internal imports

File: api/albums-router.js (hunk 3 above)
Comment (kenzable) [reply api/albums-router.js:33]: 👍

File: api/albums-controller.js [hunk 8]
```diff
@@ -133,5 +154,7 @@ module.exports = {
   getSharedAlbums,
   initializeCache,
   joinAlbum,
+  joinPendingAlbums,
   shareAlbum,
-}
+  getShareTokens,
```
Comment (kenzable) [inline api/albums-controller.js]: Doesn't look like we need to export this, yeah? Since it's coming from ./api/user.js

File: api/user.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+const db = firebase.firestore();
+
+function addShareToken(userEmail, shareToken) {
+  const userRef = db.collection('users').doc(userEmail);
```
Comment (kenzable) [inline api/user.js:5]: Does this cover the case where the user doesn't already exit?

File: api/user.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  .catch(err => {
+    console.error(err);
+    throw err;
+  });
+}
+
+function getShareTokens(userEmail) {
+  db.collection('users').doc(userEmail).get()
+  .then((user) => {
+    return user.data().shareTokens;
```
Comment (kenzable) [inline api/user.js]: Need to ensure this always returns an array, even if empty (or else update the map you use when iterating over the share tokens to handle undefined cases).

File: api/index.js (hunk 5 above)
Comment (kenzable) [reply api/index.js:34]: Do we need to create the user in firebase here if they don't already exist?

Comment (que-sara-sara) [reply api/index.js:34]: i think this is actually where we want to verify that the user is already in firebase from the admin right?

File: src/index.js (hunk 6 above)
Comment (que-sara-sara) [reply src/index.js]: hmm i think this is a rogue merge conf resolution.

File: api/index.js (hunk 5 above)
Comment (kenzable) [reply api/index.js:34]: Ah! Ok so create the user account in firebase when the admin adds them to the master list? That works 👍

File: api/albums-controller.js [hunk 11]
```diff
@@ -59,14 +62,24 @@ async function joinAlbum(userToken, userId, shareToken) {
   return result;
 }
 
+async function joinPendingAlbums(userToken, userId) {
+  const shareTokens = getShareTokens(userId);
+  const joinAllPromises = shareTokens.map(token => {
+    return async () => {
+      return joinAlbum(userToken, userId, token);
```
Comment (kenzable) [inline api/albums-controller.js:69]: If joinAlbum already returns a promise, no need to wrap this function call in another function.

**Review by kenzable — APPROVED** [review]

**Pull request merged**